server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -1695,8 +1697,20 @@ public AliasFilter buildAliasFilter(ClusterState state, String index, Set<String
     /**
      * Returns a new {@link QueryRewriteContext} with the given {@code now} provider
      */
-    public QueryRewriteContext getRewriteContext(LongSupplier nowInMillis) {
-        return new QueryRewriteContext(parserConfig, client, nowInMillis);
+    public QueryRewriteContext getRewriteContext(LongSupplier nowInMillis, IndicesRequest indicesRequest) {
+        Index[] indices = indexNameExpressionResolver.concreteIndices(clusterService.state(), indicesRequest);
+        Map<String, Set<String>> modelsForFields = new HashMap<>();
+        for (Index index : indices) {
+            Map<String, Set<String>> fieldsForModels = indexService(index).getMetadata().getFieldsForModels();
+            for (Map.Entry<String, Set<String>> entry : fieldsForModels.entrySet()) {
+                for (String fieldName : entry.getValue()) {
+                    Set<String> models = modelsForFields.computeIfAbsent(fieldName, v -> new HashSet<>());
+                    models.add(entry.getKey());
+                }
+            }
+        }
```

Comment:
Rebuilding the `modelsForFields` map like this every time is obviously inefficient, but it should be fine for a small number of `semantic_text` fields. 

I have some ideas about how to cache `modelsForFields` per index that should make this more performant, but it requires larger changes that would bloat this PR and increase its scope considerably. Probably better to iterate on this in a separate PR.